### PR TITLE
Move test generation into `FuzzEngine`

### PR DESCRIPTION
**Pull request by horsefacts**
This is a gnarly cleanup PR that primarily moves generating a fuzzed test state and converting it to orders into `FuzzEngine` and out of `FuzzMain.t.sol`.

- Move `FuzzTestContext` generation out of test body and into a `FuzzEngine#generate` lifecycle [function](https://github.com/ProjectOpenSea/seaport/blob/2ceaa8e61ad9b05cc4e7b468dd81f8c52740dd1a/test/foundry/new/helpers/FuzzEngine.sol#L95).
- Update `FuzzEngine#run` to [generate test state](https://github.com/ProjectOpenSea/seaport/blob/2ceaa8e61ad9b05cc4e7b468dd81f8c52740dd1a/test/foundry/new/helpers/FuzzEngine.sol#L67) based on fuzz inputs. 
- Update `FuzzMain.t.sol` to call the new `FuzzEngine#run`. It's [a lot cleaner](https://github.com/ProjectOpenSea/seaport/blob/horsefacts/combine-generator-test-context/test/foundry/new/FuzzMain.t.sol#L15).
- Prefix everything fuzz-related with `Fuzz`, most notably `TestContext` -> `FuzzTestContext` and `GeneratorContext` -> `FuzzGeneratorContext`. If it starts with `Fuzz`, it's fuzz test related.
- Extract `FuzzEngineLib` to its own file. (This was actually duplicated—we were using the version inline in `FuzzEngine.sol`).
- Clean up some imports, unused functions, and mutabilities.

### CODE REVIEW

**Review by DJViau — APPROVED** [review]
Wow that was a lot to parse :)

It looks good, but there were spots where big chunks were moved or deleted that were hard to parse.  Let me know if you think anything is worth scrutinizing more closely (I suspect not but I'd be happy to take a closer look if you have any anxieties).

File: test/foundry/new/FuzzEngine.t.sol [hunk 1]
```diff
@@ -206,17 +233,15 @@ contract FuzzEngineTest is FuzzEngine {
...
+        /** expectedActions[4] = seaport.cancel.selector;
+        expectedActions[5] = seaport.validate.selector; */
 
-        TestContext memory context = TestContextLib.from({
+        FuzzTestContext memory context = FuzzTestContextLib.from({
             orders: orders,
             seaport: seaport,
-            caller: address(this),
-            fuzzParams: FuzzParams({ seed: 0 })
+            caller: address(this)
```
[the remark this answers is not in the document]
Comment (horsefacts) [reply test/foundry/new/FuzzEngine.t.sol:243]: Nope, just missed one. We could omit it every time the fuzz seed is `0`, but probably better to be explicit.
